Experimental/src/org/sleuthkit/autopsy/experimental/textclassifier/TextClassifierTrainer.java

```diff
@@ -0,0 +1,256 @@
+        TrainingParameters params = new TrainingParameters();
+        params.put(TrainingParameters.CUTOFF_PARAM, Integer.toString(0));
+        params.put(TrainingParameters.ALGORITHM_PARAM, TextClassifierUtils.ALGORITHM);
+        if (oldModelPath != null) {
+            params.put("MODEL_INPUT", oldModelPath);
+        }
+
+        return DocumentCategorizerME.train(TextClassifierUtils.LANGUAGE_CODE, sampleStream, params, new DoccatFactory());
+    }
+
+    @Messages({
+        "TextClassifierTrainer.noDocs.text=No documents found. You may need to run the Ingest Module for File Type Detection",
+        "TextClassifierTrainer.fetching.text=Fetching training documents",
+        "TextClassifierTrainer.converting.text=Converting training documents",
+        "TextClassifierTrainer.needNotable.text=Training set must contain at least one notable document",
+        "TextClassifierTrainer.needNonnotable.text=Training set must contain at least one nonnotable document",})
+    /**
+     * Fetches the training data and converts it to a format OpenNLP can use.
+     *
+     * @return training data usable by OpenNLP
+     */
+    private ObjectStream<DocumentSample> processTrainingData(ReportProgressPanel progressPanel) throws TskCoreException {
+
+        progressPanel.updateStatusLabel(NbBundle.getMessage(this.getClass(), "TextClassifierTrainer.fetching.text"));
+
+        List<AbstractFile> allDocs;
+        allDocs = fetchAllDocuments();
+
+        LOGGER.log(Level.INFO, "There are {0} documents", allDocs.size());
+
+        if (allDocs.isEmpty()) {
+            progressPanel.complete(ReportStatus.ERROR);
+            progressPanel.updateStatusLabel(NbBundle.getMessage(this.getClass(), "TextClassifierTrainer.noDocs.text"));
+            throw new TskCoreException("No documents found. You may need to run the Ingest Module for File Type Detection.");
+        }
+
+        Set<Long> notableObjectIDs = fetchNotableObjectIDs();
+
+        progressPanel.updateStatusLabel(NbBundle.getMessage(this.getClass(), "TextClassifierTrainer.converting.text"));
+        progressPanel.setMaximumProgress(allDocs.size());
+        List<DocumentSample> docSamples = new ArrayList<>();
+        String label;
+        boolean containsNotableDocument = false;
+        boolean containsNonNotableDocument = false;
+        for (AbstractFile doc : allDocs) {
+            if (notableObjectIDs.contains(doc.getId())) {
+                label = TextClassifierUtils.NOTABLE_LABEL;
+                containsNotableDocument = true;
+            } else {
+                label = TextClassifierUtils.NONNOTABLE_LABEL;
+                containsNonNotableDocument = true;
+            }
+            DocumentSample docSample = new DocumentSample(label, TextClassifierUtils.extractTokens(doc));
+            docSamples.add(docSample);
+
+            progressPanel.increment();
+        }
+
+        if (!containsNotableDocument) {
```

Comment:
Why is this a requirement?  There could be cases where all documents were OK or Notable.

Reply:
This requirement is an oversimplification of the following two constraints:

1. OpenNLP will not train a model unless at least two categories occur in the training data. This does not have to mean that every batch contains both categories. 
2. IncrementalNaiveBayesTrainer.mergeInParameters() was written assuming that all of the categories ("outcomes" in OpenNLP's lingo) occurred in the first batch.

What this means is that the first batch needs to contain both classes. I'll change the code to use this requirement instead. I'll also look into whether I can get it to work even if the first batch is all of the same category.